[WIP][vcpkg] vcpkg can link wrong libraries (issue #5540 b)

Neumann-A opened the pull request:
fixes #5540 b

Feel free to change. Is just an example how to solve part of the problem. 

__Overview__
TODOs: https://github.com/Microsoft/vcpkg/pull/5543#issuecomment-480847795
OLD FIND_PACKAGE PORTS:  https://github.com/Microsoft/vcpkg/pull/5543#issuecomment-481114923
AFFECTED PORTS: https://github.com/Microsoft/vcpkg/pull/5543#issuecomment-481138549 (not updated; too many)
CMAKE_DEBUG_SUFFIX LIST: https://github.com/Microsoft/vcpkg/pull/5543#issuecomment-481228478

__Some Numbers:__
Number of packages which needed changed variables: around 60; (Thats somewhere around 6% of all packages)
Number of packages which consume those variables: around 120; (which is more than 10% of all packages)
Number of packages which actually fix the linkage issue in their target files: ???

__Status__
REMAINING ISSUES: all current regressions are errors on the package side logic to find a library and must be fixed. 
CURRENT STATUS: closed due to wrong approach

cenit commented:
another path is working with every upstream package to export proper config file and getting in contact with Kitware to update many modules to find package and deal with release/debug libraries (which should have different names also...). I tried this path, it is very long, and very very slow. Some shortcut would be extremely appreciated.

Neumann-A commented:
@cenit The problem is not only to update find_package and that libraries should export configs but also consuming libraries need to update their build scripts. I think that is something which will never 100% happen so vcpkg must support a shortcut.

the complete fix will probably look something like:
```
get_cmake_property(ALL_VARS VARIABLES)
list(FILTER ${ALL_VARS} INCLUDE REGEX "\$[${name}|${uppercasename}|${lowercasename}][0-9]*_[INC|LIB]")
foreach(PACKAGE_VAR ${ALL_VARS})
 #fix the paths here
endforeach()
unset(ALL_VARS)
```

will implement it if i have the time.

cenit commented:
@Neumann-A why do you say so? optimized;path_to_lib;debug;path_to_lib works as expected and is filled inside <PackageName>_LIBRARY by the wonderful select_library_configurations(basename) function inside a proper Find<PackageName>.cmake... so if the downstream project just links to <PackageName>_LIBRARY everything works as expected and you have config support "for free".
Anyway, I fully support you when you say that this is a way-too-long road, which will never achieve 100% completion

Neumann-A commented:
hmm can somebody get the regex correctly to also include HDF5 and possible others?
https://cmake.org/cmake/help/v3.14/module/FindHDF5.html

something like:
`"^[${name}|${uppercasename}|${lowercasename}][^INC|^LIB]_[INC|LIB]"`

cenit commented:
The procedure is to open a PR on https://gitlab.kitware.com/cmake/cmake
Unfortunately at the moment I am too busy on other tasks, otherwise I'd have helped you

Neumann-A commented:
Can somebody from the team supply me with all cmake config logs from CI after the last commit?
I would like to filter them for `Value of` to see which package variables are not properly set.

Rastaban commented:
@Neumann-A, the CI system does not keep around the logs from passing builds and the files edited are not part of the set that triggers a rebuild of ports so nothing was actually built as part of the CI test for the last commit.  Supporting this kind of thing in the test system is on my backlog, but I am not sure when I will be able to make those changes.  You can run the ci command on your own machine (`vcpkg ci [triplet]`.) It takes about 14 hours to finish on our VMs but you would have all of the logs.  If you don't want to tie up your machine for that long then let me know and I can run it on my own over the weekend.

Neumann-A commented:
TODO:
- decide if `<packagename>_LIBRARY|LIBRARIES` should always contain keywords debug and optimized even if not required to build (could solve the problem of debug/release targets being overwritten)  
- decide if `MinSizeRelease` and `RelWithDebInfo` should always be mapped to Release using `CMAKE_MAP_IMPORTED_CONFIG_<config>` if not set

Neumann-A commented:
> decide if <packagename>_LIBRARY|LIBRARIES should always contain keywords debug and optimized even if not required to build (could solve the problem of debug/release targets being overwritten)

this is required because of multi configuration generators like visual studio. Linkage of libraries for debug/release may not be correct if the cmake files link directly to either debug or release library files instead of targets or using optimized/debug.  (old find_package style)

Neumann-A commented:
note to https://github.com/Microsoft/vcpkg/pull/5625#issuecomment-480800229
1 and 2 (fix at start): fixes the multi configuration generator issue of find_package
3 and 4 (fix at end): will not fix this issue

Neumann-A commented:
@cenit: This fixes your opencv issue ... .

Lets see what CI says to this one (probably needs to be triggered manually?)

Neumann-A commented:
TODOs:
- [ ] Wait for CI regressions (if any)
- [ ] Remove/Comment out cmake messages.

Neumann-A commented:
List of ports linked wrongly (marked means fixed; List will be updated and extended when my CI runs or throws errors):
- [x] LIBLZMA
- [x] GIF
- [x] SDL

(The cmake messages really help to find those wrongly linked libraries)

Neumann-A commented:
@Rastaban: Can you create a list of ports which have indirect dependencies (meaning dependencies which dependent on the list in https://github.com/Microsoft/vcpkg/pull/5543#issuecomment-481114923; second level dependencies; don't know how to call those)? All those ports must be checked for correct linkage of its second level dependencies because the target file of the direct dependent might be wrong if the target is not manually fixed within the portfile.

Neumann-A commented:
affected libraries (incomplete list as of 231/958 (sdl ports dependencies not included due to cascade failure))
- boost-iostream
- grpc
- libevent
- libpng
- libpq
- libtheora
- libwebp
- libxml2
- opencv
- poco
- thrift
- tiff
- vtk
- yara

seanwarren commented:
@Neumann-A, I've tested this on quite a large project and it seems to do a pretty good job. I'm seeing two issues:

Building with `x64-windows-static` using visual studio I'm seeing two regressions.

**qt5** 
In the link line for the debug configuration I have:
`${VCPKG_ROOT}\installed\x64-windows-static\plugins\platforms\qwindows.lib`
where it should be: 
`${VCPKG_ROOT}\installed\x64-windows-static\debug\plugins\platforms\qwindows.lib`

**libpq**
In the link line for the debug configuration I have:
`C:\Users\CIMLab\Documents\vcpkg-pristine\installed\x64-windows-static\debug\lib\libpq.lib`
where it should be
`C:\Users\CIMLab\Documents\vcpkg-pristine\installed\x64-windows-static\debug\lib\libpqd.lib`

Neumann-A commented:
@seanwarren: Can you attach the CMake configurations logs (config-triplet-(detailed-stuff).out.log) for those failing projects? There should be lines reading `Replacing: <Something> \n with: <Someotherthing>`

seanwarren commented:
It's actually not a `vcpkg` project (just consuming `vcpkg`). In the CMake output I have

```
-- Found PostgreSQL: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib (found version "9.603")
-- All-filtered-vars:PostgreSQL_LIBRARIES;PostgreSQL_LIBRARY;PostgreSQL_LIBRARY_ADDITIONAL_SEARCH_SUFFIXES;PostgreSQL_LIBRARY_DIR;PostgreSQL_LIBRARY_DIRS;PostgreSQL_LIBRARY_DIR_MESSAGE;PostgreSQL_LIBRARY_PATH_DESCRIPTION;PostgreSQL_LIBRARY_TO_FIND
-- Value of PostgreSQL_LIBRARIES: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib
-- Replacing PostgreSQL_LIBRARIES: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib
-- with PostgreSQL_LIBRARIES: debug;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/debug/lib/libpq.lib;optimized;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib
-- Value of PostgreSQL_LIBRARY: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib
-- Replacing PostgreSQL_LIBRARY: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib
-- with PostgreSQL_LIBRARY: debug;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/debug/lib/libpq.lib;optimized;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib/libpq.lib
-- Value of PostgreSQL_LIBRARY_ADDITIONAL_SEARCH_SUFFIXES: PostgreSQL/11/lib;PostgreSQL/10/lib;PostgreSQL/9.6/lib;PostgreSQL/9.5/lib;PostgreSQL/9.4/lib;PostgreSQL/9.3/lib;PostgreSQL/9.2/lib;PostgreSQL/9.1/lib;PostgreSQL/9.0/lib;PostgreSQL/8.4/lib;PostgreSQL/8.3/lib;PostgreSQL/8.2/lib;PostgreSQL/8.1/lib;PostgreSQL/8.0/lib
-- Debug and Release values or the same! Propbably only library name without path (please check): PostgreSQL/11/lib;PostgreSQL/10/lib;PostgreSQL/9.6/lib;PostgreSQL/9.5/lib;PostgreSQL/9.4/lib;PostgreSQL/9.3/lib;PostgreSQL/9.2/lib;PostgreSQL/9.1/lib;PostgreSQL/9.0/lib;PostgreSQL/8.4/lib;PostgreSQL/8.3/lib;PostgreSQL/8.2/lib;PostgreSQL/8.1/lib;PostgreSQL/8.0/lib
-- Value of PostgreSQL_LIBRARY_DIR: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib
-- Replacing PostgreSQL_LIBRARY_DIR: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib
-- with PostgreSQL_LIBRARY_DIR: debug;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/debug/lib;optimized;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib
-- Value of PostgreSQL_LIBRARY_DIRS: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib
-- Replacing PostgreSQL_LIBRARY_DIRS: C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib
-- with PostgreSQL_LIBRARY_DIRS: debug;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/debug/lib;optimized;C:/Users/CIMLab/Documents/vcpkg-pristine/installed/x64-windows-static/lib
-- Value of PostgreSQL_LIBRARY_DIR_MESSAGE: Set the PostgreSQL_LIBRARY_DIR cmake cache entry to the top-level directory containing the PostgreSQL libraries.
-- Debug and Release values or the same! Propbably only library name without path (please check): Set the PostgreSQL_LIBRARY_DIR cmake cache entry to the top-level directory containing the PostgreSQL libraries.
-- Value of PostgreSQL_LIBRARY_PATH_DESCRIPTION: top-level directory containing the PostgreSQL libraries.
-- Debug and Release values or the same! Propbably only library name without path (please check): top-level directory containing the PostgreSQL libraries.
-- Value of PostgreSQL_LIBRARY_TO_FIND: libpq
-- Debug and Release values or the same! Propbably only library name without path (please check): libpq

```

Neumann-A commented:
@seanwarren That output suggest that the issue is somewhere else. The libpq issue could be a portbug. I am currently scanning issues which could be related to this PR

seanwarren commented:
OK fair enough, although the correct libraries are linked in master

Neumann-A commented:
@seanwarren Yeah portbug. The problem is in the CMakeLists.txt which sets `set(CMAKE_DEBUG_POSTFIX "d")`. So the debug library will be installed as `libpqd` which cannot be found by `find_package`

seanwarren commented:
It that a bug per-se? Many libraries use a `d` debug postfix. As far as I'm aware this is just a choice of convention. Changing this postfix would require patching (at least `qt5`, `soci`, `gdal` and presumably any projects using vcpkg consuming `libpq` and `libpqxx` that expect the `d` postfix

Neumann-A commented:
@seanwarren The problem here is that find_package defines the default behavior and it assumes nothing about release/debug libpq it just searches for libpq and not libpqd. 

The problem with qt5 is also a portbug in vcpkg-cmake-wrapper:
```
function(add_qt_library _target)
    foreach(_lib IN LISTS ARGN)
        find_library(${_lib}_LIBRARY_DEBUG NAMES ${_lib}d PATH_SUFFIXES plugins/platforms)
        find_library(${_lib}_LIBRARY_RELEASE NAMES ${_lib} PATH_SUFFIXES plugins/platforms)
        set_property(TARGET ${_target} APPEND PROPERTY INTERFACE_LINK_LIBRARIES 
        \$<\$<NOT:\$<CONFIG:DEBUG>>:${${_lib}_LIBRARY_RELEASE}>\$<\$<CONFIG:DEBUG>:${${_lib}_LIBRARY_DEBUG}>)
    endforeach()
endfunction()
```
here it is assumed that debug plugins have a d suffix which is apparently not the case. 

Instead of assuming behavior (like d suffixes for debug libraries) we should look into the actual behavior and take that into account (because we might not be able to change that and consuming packages might behave differently and out of vcpkgs control).

cenit commented:
Note that it is better from the cmake point of view to have libraries with different names between release and debug. 

In fact, it's exactly the opposite from what you say: for a *modern* `Find<PackageName>.cmake` file, it's almost necessary to have different names for debug and release libraries and not just a path difference. I'd strongly suggest you to **not** rename any *d debug library into one without suffix. Having reached that point is a goal often hard obtained. After this goal, there are the appropriate mechanisms inside CMake to correctly distinguish the configurations, both for *old* symbols (_LIBRARIES) and for modern targets, when consumed by downstream projects.

cenit reviewed (commented):
just some comments here and there

seanwarren commented:
>The problem with qt5 is also a portbug in vcpkg-cmake-wrapper:
You're right here actually - Qt libraries in the main folder have a `d` postfix while in the plugin folders they do not. I think this problem is  unrelated to this PR

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -15,6 +15,18 @@ if(VCPKG_TOOLCHAIN)
 endif()
 
+if(NOT DEFINED CMAKE_BUILD_TYPE) 
+#If the build type is not defined we are generating with a multi config generator
```
CMAKE_BUILD_TYPE is not defined if you generate visual studio project files consuming the vcpkg toolchain for external projects as it is suggested by vcpkg. 
That is why there is a lot of open issue related to this (Examples: #4435 #3463 and probably other mentioned within this PR)
The thing why it is included in this PR: It will cause CMake to link against Debug libraries in the case of MinSizeRel and RelWithDebInfo

Neumann-A commented on ports/sdl1/portfile.cmake:
```diff
@@ -42,5 +42,5 @@ if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
 if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
     file(MAKE_DIRECTORY ${CURRENT_PACKAGES_DIR}/debug/lib/manual-link)
-    file(RENAME ${CURRENT_PACKAGES_DIR}/debug/lib/SDLmain.lib ${CURRENT_PACKAGES_DIR}/debug/lib/manual-link/SDLmaind.lib)
+    file(RENAME ${CURRENT_PACKAGES_DIR}/debug/lib/SDLmain.lib ${CURRENT_PACKAGES_DIR}/debug/lib/manual-link/SDLmain.lib)
```
Not true for CMake users. Only true for VS integration and explicit linkage of the file. 
The problem is that this change already broke every downstream CMake user....
With the rename I fix the issue in CMake find_package and also make configurations in VS nicer since the library name is the same for Debug and Release making only one explicit change necessary (Yeah can select all configurations in VS and just add one dependency instead of two)

Neumann-A commented:
> Note that it is better from the cmake point of view to have libraries with different names between release and debug.

I would agree if the find_package call for those library would have that feature to differentiate between debug and release. Without it I am currently saying no because we have to assume that find_package must work for all users under all circumstances and adding d to the library breaks that. (Or the user is required internal knowledge of how vcpkg acts in this circumstances.)

Neumann-A commented:
Ports setting CMAKE_DEBUG_POSTFIX and probably breaking downstream find_package calls if they do not use targets or find_package does not support differentiation between debug/release: (if the package cannot be found by find_package it can be removed from the list)
- abseil
- argtable2
- aubio
- benchmark
- breakpad
- cairo
- cctz
- darknet
- dcmtk
- dlib
- flann
- freeglut
- freeimage
- freeopcua
- geos
- gflags
- glslang
- gsl
- gtest
- imgui
- jbig2dec
- jxrlib
- lcms
- libdisasm
- libgeotiff
- libidn2
- liblsl
- libmaxminddb
- libmspack
- libmupdf
- libpopt
- libraw
- librtmp
- libwebp
- lz4
- open62541
- pixman
- plib
- pthreads
- realsense2
- rocksdb
- snappy
- speex
- suitesparse
- ~~tiff~~ has target
- zstd
- zziplib

Confirmed to break downstream find_package
- libpq
- libpqxx

Neumann-A commented:
[ ] fix PCRE handling within ports. They tend to do it all a bit different resulting in the wrong library being linked in either debug or release (depends on the handling....)

seanwarren commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -207,31 +221,70 @@ macro(find_package name)
+                    endif()
+                    if(NOT "${_pkg_var_debug}" STREQUAL "${_pkg_var_release}")
+                        message(STATUS "Replacing ${_pkg_var}: ${${_pkg_var}}")
+                        set(${_pkg_var} "debug;${_pkg_var_debug};optimized;${_pkg_var_release}")
```
@Neumann-A - could you create a list of files in the lib directory and cache it at the start of the process?

Neumann-A replied:
@seanwarren I like that idea. Probably will implement this because having a check if the file really exist should always be done. (But I found searching for the file again too complicated -> to much extra code to maintain)
Currently only 5 ports are confirmed to have the wrong linkage behavior due to the vcpkg added extra "d". As such I think vcpkg should remove the extra "d" in those ports and simply follow the priority list suggested here https://github.com/Microsoft/vcpkg/issues/6014#issuecomment-481548710

seanwarren replied:
@Neumann-A - sounds like a reasonable approach to me

philjdf commented:
Slightly related comments...

It would be great if the vcpkg documentation was better (e.g. what is the best practice output of a port? What should it expect the dependencies to do?), so it would allow people like me who aren't CMake experts to make higher-quality contributions, and learn stuff in the process.

I've updated the vcpkg Arrow port two or three times. I'm far from a CMake expert, so it feels a bit like stabbing in the dark, but I've found that I've always needed some FindX.cmake files in the Arrow port, and I've also needed FindX.cmake downstream when I use the vcpkg Arrow, to pick up all the correct debug/release libs.

I've noticed libs may or may not append 'd' for debug builds, and may or may not append '_static' for static triplets. Is this because they're trying to be consistent with the CMake built-in find_package?

Neumann-A commented:
@philjdf I am also not an CMake expert but I have some expectations concerning CMake and one of them is that find_package(<packagename>) should always do the right thing if used downstream and the package is within vcpkg. As such there should be no need for any FindX.cmake in downstream targets. If there is need for FindX.cmake in a downstream target then vcpkg should probably fix X or let X  provide a FindX.cmake itself within vcpkg. 

> what is the best practice output of a port? 

Depends on what the package defines. But in easy words I would say provided either cmake targets or at least be correctly find able using find_package in a downstream project. 

> I've noticed libs may or may not append 'd' for debug builds, and may or may not append '_static' for static triplets. Is this because they're trying to be consistent with the CMake built-in find_package?

I would say, this seems to be a c++ historic issue because c++ people "were too self centered" to come up with an appropriate general solution earlier which had general consent (like more than 20 years ago). But lets not go down that rabbit hole and just deal with those current "separate islands of general consent" and build bridges between them, so that we can simply build a bigger island of general consent :). My general approach for library naming, if i would invent a new one, would probably be (lib)packagename-version_(r|d)(libsuffix). "(lib)" if static (as it is done on linux as far as i know). "(r|d)" to differentiate between debug and release and deliver with a proper cmakelists.txt which exports the library with targets/configs. Because modern CMake is all about targets and library naming is not that important at all ;)

Neumann-A commented:
@vicroms @Rastaban Is there a vcpkg debug mode which sets a CMake variable? Something like VCPKG_DEBUG_MODE. So that I can hide the messages in normal mode. Also something like VCPKG_DEBUG_MODE_STRICT would probably be helpful to make some of the debug messages FATAL_ERRORS.

Neumann-A commented:
I will add some three more minor changes after CI has run:
includes: 
- remove second setting of CMAKE_DEBUG_POSTFIX"d" in libpq
- remove "d" from libpopt
- more debug messages in find_package to make sure everything works as expected (would be great if CI could keep the buildtree *.logs around because those are needed for the analysis if everything works as expected)

Neumann-A commented:
@vicroms: Can I safely push new commits if CI is once finished or does this delete the logs of last CI run? I would be interested in any unexpected regressions caused by this pr because my new commits will probably not solve those. 
And somehow this PR got larger than I ever expected it to be.

Neumann-A commented:
Seems like having `<package>_RELEASE` and `<package>_DEBUG` is not actually enough for knowing if those libraries are correctly set. See: (from netcdf-c)

```
-- VCPKG-find_package value of SZIP_BIN_SEARCH_DIRS: /bin;/usr/bin
-- VCPKG-find_package value of SZIP_FOUND: 1
-- VCPKG-find_package value of SZIP_INCLUDE_DIR: C:/Scratch/neumann/vcpkg/installed/x64-windows/include
-- VCPKG-find_package value of SZIP_INCLUDE_DIRS: C:/Scratch/neumann/vcpkg/installed/x64-windows/include
-- VCPKG-find_package value of SZIP_INCLUDE_SEARCH_DIRS: /include;/include/szip;/usr/include;/usr/include/szip
-- VCPKG-find_package value of SZIP_LIBRARIES: optimized;C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib;debug;C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib
-- VCPKG-find_package value of SZIP_LIBRARY: optimized;C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib;debug;C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib
-- VCPKG-find_package value of SZIP_LIBRARY_DEBUG: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib
-- VCPKG-find_package value of SZIP_LIBRARY_DEBUG: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib
-- VCPKG-find_package value of SZIP_LIBRARY_PATH: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib
-- VCPKG-find_package value of SZIP_LIBRARY_RELEASE: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/szip.lib
-- VCPKG-find_package value of SZIP_LIB_DIR: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib
-- VCPKG-find_package value of SZIP_LIB_SEARCH_DIRS: /lib;/usr/lib
-- VCPKG-find_package value of SZIP_SEARCH_DEBUG_NAMES: sz_d;libsz_d
-- VCPKG-find_package value of SZIP_SEARCH_RELEASE_NAMES: sz;libsz;szip
```

```
-- VCPKG-find_package value of CURL_FOUND: TRUE
-- VCPKG-find_package value of CURL_INCLUDE_DIR: C:/Scratch/neumann/vcpkg/installed/x64-windows/include
-- VCPKG-find_package value of CURL_INCLUDE_DIRS: C:/Scratch/neumann/vcpkg/installed/x64-windows/include
-- VCPKG-find_package value of CURL_LIBRARIES: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/libcurl.lib
-- VCPKG-find_package value of CURL_LIBRARY: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/libcurl.lib
-- VCPKG-find_package value of CURL_LIBRARY_DEBUG: CURL_LIBRARY_DEBUG-NOTFOUND
-- VCPKG-find_package value of CURL_LIBRARY_RELEASE: C:/Scratch/neumann/vcpkg/installed/x64-windows/lib/libcurl.lib
-- VCPKG-find_package value of CURL_VERSION_STRING: 7.61.1-DEV
```

The libcurl line `-- VCPKG-find_package value of CURL_LIBRARY_DEBUG: CURL_LIBRARY_DEBUG-NOTFOUND` will generate the wrong target information for netcdf-c without the required debug information

vicroms commented:
@Neumann-A 

As long as the CI build has finished is safe to push new commits. We can retrieve logs for the old runs without problems.

But, if the CI build is still running and you push new commits, the ongoing CI build is cancelled and a new CI build for the latest commit is started.

vicroms commented:
### [Failure logs for build 2579603](https://github.com/Microsoft/vcpkg/files/3081597/failureLogs-5543-2579603.zip)

| x64-windows  | master | test | notes          |
| ------------ | ------ | ---- | -------------- |
| sfml         | Pass   | Fail | **Regression** |
| openmvg      | Pass   | Fail | **Regression** |
| arrow        | Pass   | Fail | **Regression** |
| magnum       | Pass   | Fail | **Regression** |
| cartographer | Pass   | Fail | **Regression** |
| sophus       | Pass   | Fail | **Regression** |
| cairomm      | Pass   | Fail | **Regression** |
| libfreenect2 | Pass   | Fail | **Regression** |
| vxl          | Pass   | Fail | **Regression** |
| libgd        | Pass   | Fail | **Regression** |
| wxwidgets    | Pass   | Fail | **Regression** |
| ogre         | Pass   | Fail | **Regression** |

| x64-windows-static | master | test | notes          |
| ------------------ | ------ | ---- | -------------- |
| arrow              | Pass   | Fail | **Regression** |
| fizz               | Pass   | Fail | **Regression** |
| magnum             | Pass   | Fail | **Regression** |
| libfreenect2       | Pass   | Fail | **Regression** |
| vxl                | Pass   | Fail | **Regression** |
| wxwidgets          | Pass   | Fail | **Regression** |
| openmvg            | Pass   | Fail | **Regression** |

| x64-osx | master | test | notes          |
| ------- | ------ | ---- | -------------- |
| caffe2  | Pass   | Fail | **Regression** |
| magnum  | Pass   | Fail | **Regression** |
| podofo  | Pass   | Fail | **Regression** |
| opencv  | Pass   | Fail | **Regression** |

| arm64-windows | master | test | notes          |
| ------------- | ------ | ---- | -------------- |
| sophus        | Pass   | Fail | **Regression** |

| x86-windows  | master | test | notes          |
| ------------ | ------ | ---- | -------------- |
| sfml         | Pass   | Fail | **Regression** |
| magnum       | Pass   | Fail | **Regression** |
| cartographer | Pass   | Fail | **Regression** |
| sophus       | Pass   | Fail | **Regression** |
| cairomm      | Pass   | Fail | **Regression** |
| libfreenect2 | Pass   | Fail | **Regression** |
| vxl          | Pass   | Fail | **Regression** |
| libgd        | Pass   | Fail | **Regression** |
| wxwidgets    | Pass   | Fail | **Regression** |
| ogre         | Pass   | Fail | **Regression** |
| openmvg      | Pass   | Fail | **Regression** |

| x64-linux         | master | test | notes          |
| ----------------- | ------ | ---- | -------------- |
| sqlite3           | Pass   | Fail | **Regression** |
| libogg            | Pass   | Fail | **Regression** |
| openal-soft       | Pass   | Fail | **Regression** |
| lcms              | Pass   | Fail | **Regression** |
| shapelib          | Pass   | Fail | **Regression** |
| freeglut          | Pass   | Fail | **Regression** |
| qhull             | Pass   | Fail | **Regression** |
| rocksdb           | Pass   | Fail | **Regression** |
| libqrencode       | Pass   | Fail | **Regression** |
| glib              | Pass   | Fail | **Regression** |
| boost-locale      | Pass   | Fail | **Regression** |
| liblemon          | Pass   | Fail | **Regression** |
| brotli            | Pass   | Fail | **Regression** |
| expat             | Pass   | Fail | **Regression** |
| snappy            | Pass   | Fail | **Regression** |
| minizip           | Pass   | Fail | **Regression** |
| lz4               | Pass   | Fail | **Regression** |
| liblzma           | Pass   | Fail | **Regression** |
| openssl-unix      | Pass   | Fail | **Regression** |
| openblas          | Pass   | Fail | **Regression** |
| double-conversion | Pass   | Fail | **Regression** |
| cfitsio           | Pass   | Fail | **Regression** |
| uriparser         | Pass   | Fail | **Regression** |
| glog              | Pass   | Fail | **Regression** |
| glew              | Pass   | Fail | **Regression** |
| jsoncpp           | Pass   | Fail | **Regression** |
| civetweb          | Pass   | Fail | **Regression** |

Neumann-A commented:
@vicroms: Thx for the logs. Is there a possibility to check if the linux build actually used the right git commit? The error pattern I see there (CMake syntax error due to missing ") should have been solved by commit https://github.com/Microsoft/vcpkg/pull/5543/commits/82316bd474ced7a79a79f6f5da9ac61f199cf8c5 and I would have expected the other triplets to also fail with a cmake syntax error. But thanks to the logs I can finally fix the remaining outstanding issues. I there a trick to not have `vcpkg ci triplet` to crash at the end and actually show the regressions?

ToDos:
- [ ] Fix vcpkg_cmake_wrapper for TIFF target. 
- [ ] Fix wrong glog information in CeresConfig.cmake (using import_prefix for some reason which is not set.) 
- [ ] Fix strange linkage in some packages linking to debug.lib and optimized.lib (e.g. libfreenect2)
- [x] solve conflicts by rebasing. (not yet forced pushed)
- [x] rewrite logic to check _RELEASE and _DEBUG variables for correct; Ignore any correction if a XConfig.cmake exists. 
- [ ] (necessary?) Fix CMake syntax error on linux?

Neumann-A commented:
@vicroms I tried reproducing some of the linux regressions and failed to do so. I think there was something wrong with the linux CI build on that run.


__Known remaining windows regressions :__
- [ ] magnum
- [ ] fizz (mainly due to port update of folly & fizz, will revert that change, and fix the old packages instead of the updated ones)
- [ ] recheck vxl (probably false negative)

Neumann-A commented:
@vicroms Can I get an update on the regressions?

Rastaban commented:
resolved conflict so PR test could run.

Neumann-A commented:
@Rastaban @vicroms How is the CI/regressions status?

Rastaban commented:
The only regressions in the PR system I see is caffe2 on Mac/Linux.

cenit commented:
there are many bumped packages with apparent no reason

cenit commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -219,34 +233,224 @@ macro(find_package name)
+        STRING(REPLACE "*" "\\*" _pkg_names_rgx "${_pkg_names_rgx}")
+        set(_pkg_filter_rgx "^(${_pkg_names_rgx})([^_]*_)+")
+        list(FILTER _pkg_all_vars INCLUDE REGEX ${_pkg_filter_rgx})
+        message(STATUS "VCPKG-all-package-defined-vars: ${_pkg_all_vars}")
```
these messages can pollute logs a lot. Are they necessary?

cenit commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -219,34 +233,224 @@ macro(find_package name)
+        list(FILTER _pkg_all_vars INCLUDE REGEX ${_pkg_filter_rgx})
+        message(STATUS "VCPKG-all-package-defined-vars: ${_pkg_all_vars}")
+        foreach(_pkg_var ${_pkg_all_vars})
+            message(STATUS "VCPKG-find_package value of ${_pkg_var}: ${${_pkg_var}}")
```
same as before

cenit commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -219,34 +233,224 @@ macro(find_package name)
+
+        #Filtering for variables which are probably library variables for the package.
+        list(FILTER _pkg_all_vars INCLUDE REGEX ${_pkg_filter_rgx})
+        message(STATUS "VCPKG-find_package: all-filtered-library-vars: ${_pkg_all_vars}")
```
same as before

cenit commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -219,34 +233,224 @@ macro(find_package name)
     endif()
 endmacro()
 
+#macro(find_library name)
```
remove unnecessary left-overs

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -219,34 +233,224 @@ macro(find_package name)
+                            if(NOT EXISTS "${_pkg_maybe_loc}")
+                                message(WARNING "VCPKG-Warning-find_package: Could not locate debug libraries at: ${_pkg_maybe_loc}. Trying harder")
+                                #Trying harder: (unfortunally this does not work in the _RELEASE case. )
+                                get_filename_component(_pkg_maybe_name "${${_pkg_var}}" NAME_WLE)
```
NAME_WLE only exist in newer CMake versions. Does somebody have a backwards compatible version to extract the filename without extension in CMake?

Neumann-A commented:
@Rastaban thx. I will have a look at caffe2 on linux. At least I now know the approach seems to be working for allmost all ports. The logic is still not 100% finished because there is more which could be checked.

@cenit
> there are many bumped packages with apparent no reason.

This is due to fact that I change behavior in vcpkg.cmake and this on the other hand changes (corrects) behavior of those ports. It only becomes apparent if you search the buildtree for `VCPKG-Warning-find_package`, `VCPKG-find_package`, or `VCPKG-find_package: Replacing`
Theoretically I should push the control of each port mentioned here: https://github.com/Microsoft/vcpkg/pull/5543#issuecomment-481114923 because thats the list I extracted from a local `vcpkg ci` build by searching for `VCPKG-find_package: Replacing`

Neumann-A commented on ports/ceres/portfile.cmake:
```diff
@@ -69,10 +69,11 @@ endif()
-               "get_filename_component(CURRENT_ROOT_INSTALL_DIR\n    \${CERES_CURRENT_CONFIG_DIR}/../../" CERES_TARGETS "${CERES_TARGETS}")
-file(WRITE ${CURRENT_PACKAGES_DIR}/share/ceres/CeresConfig.cmake "${CERES_TARGETS}")
+               "get_filename_component(CURRENT_ROOT_INSTALL_DIR\n    \${CERES_CURRENT_CONFIG_DIR}/../../" CERES_CONFIG "${CERES_CONFIG}")
+string(REGEX REPLACE "get_filename_component\\(CERES_BUILD_GLOG_LIBRARY_DIR" "set\(CERES_BUILD_GLOG_LIBRARY_DIR" CERES_CONFIG "${CERES_CONFIG}") # Import prefix not defined; The hints in the config are actually not required and could be removed
```
ah I see you just patched all the unnecessary behavior within the ceres config away.

Neumann-A commented on ports/folly/FindLibsodium.cmake:
```diff
@@ -0,0 +1,274 @@
+# Written in 2016 by Henrik Steffen Gaßmann <henrik@gassmann.onl>
```
file is not used. will remove it.

Neumann-A commented on ports/tiff/vcpkg-cmake-wrapper.cmake:
```diff
@@ -3,8 +3,9 @@ _find_package(${ARGS})
+
 if(TARGET TIFF::TIFF)
-  set_property(TARGET TIFF::TIFF APPEND PROPERTY INTERFACE_LINK_LIBRARIES ${LIBLZMA_LIBRARIES} JPEG::JPEG ZLIB::ZLIB)
+  set_property(TARGET TIFF::TIFF APPEND PROPERTY INTERFACE_LINK_LIBRARIES ${LIBLZMA_LIBRARIES} JPEG::JPEG ZLIB::ZLIB) # Should be fixed upstream
```
will be removed. The stroy behind the changes:
Tried setting: `LIBLZMA_LIBRARIES` to include `debug/optimized` keywords: breaks `set_property`
`debug/optimized` are only allowed within `target_link_libraries`
Changed it to `target_link_libraries` instead of `set_property` breaks other ports because `target_link_libraries` is only allowed if the target is build. 
Next try: Generator expressions as found in auto generated targets, e.g. `$<Config:DEBUG> somelib; $<Not:$<Config:DEBUG> debugsomelib`  -> Breaks other cmakestuff if the somelib actually contains a list of libraries. A possible fix would have been to use extra quotes around the variables but that requires a lot of downstream patches.
So the final solution was to put the generator expression in the paths of the libraries which is the current implementation within this pr (vcpkg.cmake)

cenit commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -339,14 +327,16 @@ macro(find_package name)
             #list(LENGTH _pkg_tmp_vars _pkg_tmp_length)
             foreach(_pkg_var ${_pkg_tmp_vars})
-                if(NOT "${${_pkg_var}}" MATCHES "NOTFOUND") #TODO does not work as expected
+                if(NOT "${${_pkg_var}}" MATCHES "NOTFOUND") #TODO: Check why this sometimes does not work as expected
```
is it possible that the double expansion fails somehow? in that case usually prepending a char helps, something like
```
if(NOT "x${${_pkg_var}}" MATCHES "xNOTFOUND")
```

Neumann-A committed:
changed toolchain file to correct find_package and find_library
moved function/macro defintions in extra cmake files.

Neumann-A committed:
removed d from zstd

Neumann-A committed:
removed extra d from lz4

Neumann-A committed:
removed extra d from libwebp

Neumann-A committed:
removed extra d from libpq

Neumann-A committed:
removed extra d from libpqxx

Neumann-A committed:
removed extra d from libopt

Neumann-A committed:
removed extra d from lcms

Neumann-A committed:
removed extra d from freeimage

Neumann-A committed:
fix arrow regression due to the d removals

Neumann-A committed:
remove extra d from snappy

Neumann-A committed:
fix qt 5 regression due to d removal in libpq

Neumann-A committed:
fix libarchive regression due to removal of extra d in lz4

Neumann-A committed:
bump gsl control due to changes in vcpkg.cmake

Neumann-A committed:
refactored ceres portfile minimal

Neumann-A committed:
fix rocksdb regression due to d removal

Neumann-A committed:
removed extra d from SDL

Neumann-A committed:
fix poco find pcre

Neumann-A committed:
fix caffe2 regression

Neumann-A committed:
fix libmysql regression due to extra d removal

Neumann-A committed:
fix libraw regression due to extra d removal

Neumann-A committed:
fix libfreenect2 regression

Neumann-A committed:
fix folly regression

Neumann-A committed:
fix fizz regression

Neumann-A committed:
fix gdal regression due to d removal

Neumann-A committed:
fix openssl linkage due to cmakes FindOpenSSL behavior

Neumann-A committed:
fix orc regression due to d removal

Neumann-A committed:
fix vtk regression due to removal of d in lz4

Neumann-A commented:
So after thinking about details a lot there seems to be a fundamental flaw in vcpkg.cmake and it can be found here:
https://github.com/microsoft/vcpkg/blob/9238e93cc32555ad7c5239376d31796e949e16da/scripts/buildsystems/vcpkg.cmake#L110
This line and the following simply does not work well with find_package/library etc. and multi configuration generators because the debug path is always put before the release path. Single configuration generators should be fine with the code. 

What has to be done for a multi configuration generator?:
- [ ]  point CMAKE_FIND_ROOT_PATH directly to the share folder -> makes it easily possible for find_package to find  package_config.cmake
- [ ] manually fix find_library calls to include the right directory if the variable to set is called x_RELEASE or x_DEBUG
- [ ] manually fix find_library calls with variable ending on LIBRARY/LIBRARIES to include a generator expression for the correct path. 
- [ ] fix regressions caused by vcpkg_cmake_wrapper.cmake which can not correctly handle debug/optimized keywords.

Neumann-A committed:
refactor toolchain file a bit

Neumann-A committed:
more refactoring. Main work should now be done by find_library

Neumann-A committed:
fixing syntax errors

Neumann-A committed:
fix syntax errors

Neumann-A committed:
fix offset in insert

Neumann-A committed:
needed policy

Neumann-A committed:
fix gdal regression due to d removal of webp

Neumann-A committed:
added find_package wrapper for pcre

Neumann-A committed:
changed pcre lookup in glib

Neumann-A committed:
Jasper needs d debug suffix due to cmakes find_package

Neumann-A committed:
remove gdal rename

Neumann-A committed:
fix problems with setting variables and the policy stack

Neumann-A committed:
fix syntax

Neumann-A committed:
more safeguarding

Neumann-A committed:
trying to fix osg

Neumann-A committed:
more syntax logic corrections.

Neumann-A committed:
add pcre debug suffix because a lot of find modules assume it

Neumann-A committed:
fix yara jansson linkage

Neumann-A committed:
fix butteraugli linkage

Neumann-A committed:
add linkage of turbojpeg back

Neumann-A committed:
fix fontconfig regression

Neumann-A committed:
install bc6h, bc7, squish libraries of nvtt and add debug postfix.
This is required by OSG to find those libraries.

Neumann-A committed:
added checks for imported location.

Neumann-A committed:
Revert "fix fontconfig regression"

This reverts commit 36975c1bfbeab9939421b39b658eec0d314daa4d.

Neumann-A committed:
bit of refactoring

Neumann-A committed:
extending the logic to search for renamed debug libraries.
current problem: _set_target_properties called with incorrect number of elements if one element is empty

Neumann-A committed:
fixed the bug in set_target_properties with empty properties

Neumann-A committed:
uncomment the debug messages

Neumann-A committed:
heavy refactoring.
renamed all functions overrides to a vcpkg_functionname
added an option to disable the override
added an guard to detect infinit recursion if an external toolchain also
overwrittes the functions

Neumann-A committed:
fix logic/syntax issues

Neumann-A committed:
fix syntax/logic due to refactoring

Neumann-A committed:
added a few new options to descalate/escalted errors/warnings

Neumann-A committed:
refactoring find_package

Neumann-A committed:
remove gdal rename

Neumann-A committed:
refined recursion guard and reworked find_library a bit.

Neumann-A committed:
added message back in

Neumann-A committed:
renamed variable

Neumann-A committed:
fix retry in find_library

Neumann-A committed:
added checking of IMPORTED_IMPLIB property and fixed a copy and paste error

Neumann-A committed:
added workaround for release executables in IMPORTED_LOCATION_DEBUG

Neumann-A committed:
better description

Neumann-A committed:
remove ogg discovery in libvorbis portfile.
Should better be done by a patch to libvorbis itself and the used find_library call

Neumann-A committed:
fix ogg search names in libtheora

Neumann-A committed:
[libvorbis] fix ogg find_library call #6558

Neumann-A committed:
fix wt regressions due to openssl change.
Will remove as soon as the OpenSSL rename is merged in master

Neumann-A committed:
[podofo] remove unnecessary find modules.

Neumann-A committed:
[apr-util] fix logic.
Checking existance of file is performed and ensured by find_library.

Neumann-A committed:
[Wavpack] Link against OpenSSL target.
Generator expression cannot deal correctly with a list.
(e.g. debug;lib_debug;optimze;lib or requires extra quotes "${Var}")

Neumann-A committed:
fix qt5 file exist logic and rename LIBRARY to LIBRARY_RELEASE

Neumann-A committed:
simplified regex

Neumann-A committed:
Fixed local scope variable shadowing parent scope variable.

Neumann-A committed:
added possibility to find debug libraries if they are missing an expected debug suffix
(main offender QT5 expecting OpenGL32d)

Neumann-A committed:
first set property overwrite commit

Neumann-A committed:
remove gflags patch

Neumann-A committed:
fixed copy paste error.

Neumann-A committed:
fix qt5 base patch.

Neumann-A committed:
some guard improvements

Neumann-A committed:
removed qt5 patch. no longer required by new upstream qt5-base

Neumann-A committed:
[smpeg2] improve SDL2 linkage

Neumann-A committed:
[signalrclient] fix cpprestsdk linkage

Neumann-A committed:
[folly] removed unnecessary patches

Neumann-A committed:
fix regex for debug retry search

Neumann-A committed:
[arrow]fix zstd linkage

Neumann-A committed:
[freerdp] remove outdated openssl module

Neumann-A committed:
[libgeotiff] fix missing debug configs

Neumann-A committed:
[geotiff] fix configs/targets

Neumann-A committed:
[pdal] fix errors; remove outdated modules

Neumann-A committed:
[libmysql] Fix CMakes integrated FindICU module
(https://gitlab.kitware.com/cmake/cmake/issues/19302)

Neumann-A committed:
[OpenEXR] Fix typo in find module and added version

Neumann-A committed:
[libmysql] fix paths in module

Neumann-A committed:
[openexr] fix version string. Would be better if we would use the find module supplied by openexr instead and patch it to the correct behavior

Neumann-A committed:
[openimageio] Use OpenEXR find module instead the integrated one.

Neumann-A committed:
searching for debug libs also requires to change the search paths to the debug libs

Neumann-A committed:
move logic from vcpkg_set_target_properties to vcpkg_set_property

Neumann-A committed:
fix implementation details

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

# Conflicts:
#	ports/libgeotiff/portfile.cmake
#	scripts/buildsystems/vcpkg.cmake

Neumann-A committed:
actually add the set property override

Neumann-A committed:
add policy to properly check for keywords in arguments

Neumann-A committed:
fix logic. Must forward first and then check the parameters or it is required to refactor the whole logic.

Neumann-A committed:
remove extra x

Neumann-A committed:
fix find library and added a bit more documentation

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

# Conflicts:
#	ports/libgeotiff/portfile.cmake
#	ports/pdal/portfile.cmake

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

# Conflicts:
#	ports/ceres/CONTROL
#	ports/ceres/portfile.cmake
#	ports/gdal/portfile.cmake

Neumann-A committed:
make finding _DEBUG libs in release non fatal and retry to search without debug suffix

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

Neumann-A committed:
fix arrow zstd linkage

Neumann-A committed:
fix poco include dir

Neumann-A committed:
reenable ninja

Neumann-A committed:
fix openexr find module for openimageio

Neumann-A committed:
find_library patch. variable was not set into parent scope

Neumann-A committed:
Merge branch 'patch-2' of https://github.com/Neumann-A/vcpkg into patch-2

Neumann-A committed:
Merge remote-tracking branch 'origin/patch-2' into patch-2

Neumann-A committed:
sometimes cmake_parse_arguments PARSE_ARGV fails so retry with other version.
names position is sometimes not the second one.
in these cases additional unused arguments are passe to find_library.
Due to that the NAMES position is variable and must be found.

Neumann-A committed:
moved some code around. Code was actually required by add_executable

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

# Conflicts:
#	ports/glib/CONTROL
#	ports/libraw/CONTROL
#	scripts/buildsystems/vcpkg.cmake

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

Neumann-A committed:
bump control to force rebuild and ci cache refresh.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

# Conflicts:
#	ports/fizz/CONTROL
#	ports/folly/CONTROL
#	ports/libmysql/CONTROL
#	ports/zstd/CONTROL

Neumann-A commented:
documenting another approach:
- do not fix find_library -> fix target_link_libraries instead
- requires correct imported targets defined in findModule.cmake (will require set_property fix from this PR)

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

Neumann-A committed:
fix find_library argument parsing.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into patch-2

# Conflicts:
#	ports/vtk/CONTROL

Neumann-A committed:
bump control of regressing ports to force a rebuild of those ports.
CI logs told me they were not even rebuild the last time

Neumann-A committed:
fix openssl target

Neumann-A committed:
bump openssl control

Fleischner commented:
pybind11 installed cmake targets look suspicious. There seems to be a quite complicated machinery that searches for an python.exe and once found it searches for the python libraries relative to this python.exe.

However vcpkg port python3 currently only builds python.dlls. So pybind11 will certainly find completely wrong .lib files in this case. 
Also it seems debug libraries are not handled.

It should just use the python libraries of the python3 port instead

angelmixu commented:
Hi @Neumann-A ! Thanks for all the work!
What happened with this PR? Has it been merged, or are you working on a different approach?

Neumann-A commented:
@angelmixu. 
It requires a different approach. It is not possibly to simply inject the variable returned by find_library with a generator expression because of:
Antipattern: `if(EXISTS LIBRARY_VAR) ` after a find_library call
Usage of `CMAKE_REQUIRED_LIBRARIES` for `CheckFunctionExists` not working with generator expressions and small other minor blockers.

So an alternativ approach is necessary with the following attributes:

fix find_library calls with variables:
LIBRARY_VAR -> point to release
LIBRARY_VAR_RELEASE -> point to release
LIBRARY_VAR_DEBUG ->point to debug
generate VCPKG_LIB_NAMES_libname_VAR with VCPKGPATH/<libname>.lib and set it to names used in the find_library call. (is properly not required but it is good to have the search names stored somewhere)
generate VCPKG_LIB_GENEXP_libname_VAR generator expression with the correct paths to release/debug

fix set_property on targets and target_set_properties:
IMPORTED_LOCATION -> Generate IMPORTED_LOCATION_RELEASE and IMPORTED_LOCATION_DEBUG (properly requires the searchnames to do an extra find_library call in release or debug folder)
IMPORTED_LOCATION_RELEASE -> check if release path
IMPORTED_LOCATION_DEBUG  -> check if debug path

fix target_link_libraries and link_libraries:
Check every library which is linked having the vcpkg base path and replace it with VCPKG_LIB_GENEXP_libname_VAR if necessary (e.g. if debug/optimized keywords are not used)

So I have a plan but no time to do it currently
